torchx/schedulers/ray_scheduler.py

```diff
@@ -0,0 +1,299 @@
+        #
+        # 2. Copy `cfg.scripts` or their directories to `~/<cfg.app_id>` on the
+        #    cluster using `ray rsync-up` if requested by `cfg.copy_scripts` or
+        #    `cfg.copy_script_dirs`.
+        #
+        # 3. Dump `cfg.actors` to a JSON file (e.g. `/tmp/<cfg.app_id>-actors.json`)
+        #    and copy to `~/<cfg.app_id>` on the cluster using `ray rsync-up`.
+        #
+        # 4. Copy `ray_main.py` which handles the actual job execution to
+        #    `~/<cfg.app_id>` on the cluster using `ray rsync-up`.
+        #
+        # 5. Start `ray_main.py` as a background process on the head node using
+        #    `ray exec`.
+
+        return cfg.app_id
+
+    def _submit_dryrun(self, app: AppDef, cfg: RunConfig) -> AppDryRunInfo[RayJob]:
+        app_id = make_unique(app.name)
+
+        cluster_cfg = cfg.get("cluster_config_file")
+        if not isinstance(cluster_cfg, str) or not os.path.isfile(cluster_cfg):
+            raise ValueError("The cluster configuration file must be a YAML file.")
+
+        job = RayJob(app_id, cluster_cfg)
+
+        def set_job_attr(cfg_name: str, cfg_type: Type) -> None:
+            cfg_value = cfg.get(cfg_name)
+            if cfg_value is None:
+                return
+
+            if not isinstance(cfg_value, cfg_type):
+                raise TypeError(
+                    f"The configuration value '{cfg_name}' must be of type {cfg_type.__name__}."
+                )
+
+            setattr(job, cfg_name, cfg_value)
+
+        set_job_attr("cluster_name", str)
+        set_job_attr("start_cluster", bool)
+        set_job_attr("copy_scripts", bool)
+        set_job_attr("copy_script_dirs", bool)
+        set_job_attr("verbose", bool)
+
+        warn_unused_copy_option = True
+
+        for role in app.roles:
+            # Replace the ${img_root}, ${app_id}, and ${replica_id} placeholders
+            # in arguments and environment variables.
+            role = macros.Values(
+                img_root=role.image, app_id=app_id, replica_id="${rank}"
+            ).apply(role)
+
+            actor = RayActor(
+                name=role.name,
+                command=" ".join([role.entrypoint] + role.args),
+                env=role.env,
+                num_replicas=role.num_replicas,
+                num_cpus=role.resource.cpu,
+                num_gpus=role.resource.gpu,
+                memory_size=role.resource.memMB,
+            )
+
+            job.actors.append(actor)
+
+            if job.copy_scripts or job.copy_script_dirs:
+                # We support copying scripts to the cluster only for DDP jobs.
+                if (
+                    # fmt: off
+                    actor.command.find(" -m torch.distributed.run ") > 1 or
+                    actor.command.find(" -m torch.distributed.launch ") > 1 or
+                    actor.command.startswith("torchrun")
+                    # fmt: on
+                ):
```

Comment:
might be nice to be able to support arbitrary `python foo.py` style commands

Reply:
Do you do the same in the rest of `torchX` as in make it easy to run arbitrary Python code which may or may not include Pytorch related code